lib/ddtrace/contrib/rack/patcher.rb

```diff
@@ -0,0 +1,58 @@
+module Datadog
+  module Contrib
+    module Rack
+      module Patcher
+        include Base
+        register_as :rack
+        option :tracer, default: Datadog.tracer
+        option :distributed_tracing, default: false
+        option :middleware_names, default: false
+        option :application
+        option :service_name, default: 'rack', depends_on: [:tracer] do |value|
+          get_option(:tracer).set_service_info(value, 'rack', Ext::AppTypes::WEB)
+          value
+        end
+
+        module_function
+
+        def patch
+          return true if patched?
+
+          require_relative 'middlewares'
+          @patched = true
+
+          return unless get_option(:middleware_names)
+
+          top = get_option(:application) || rails_app
+          retain_middleware_name(top)
+        end
+
+        def patched?
+          return @patched if defined?(@patched)
```

Comment:
Not sure I follow this... "Return patched if patched defined, otherwise return... patched"? Redundant?

Reply:
This is for having a memoized value that can be `false`.

Ideally we should have:

```rb
def patched?
  @patched ||= false
end
```
But that obviously doesn't work for *falsey* values.